### PR TITLE
Update cert-manager tutorial with HTTPProxy details

### LINKED ISSUE #1140: Add docs on how to use CertManager with HTTPProxy
**Describe the solution you'd like**
Docs exist today on how to setup CertManager to create Let's Encrypt certificates with Kubernetes Ingress resources. It would be great to also demonstrate how to accomplish this with IngressRoute objects. 

We cannot use annotations with this style, however, we can create Cert Manager "Certificate" CRD's to also let CertManager help with the certificate request flow.

### LINKED ISSUE #1570: cert-manager tutorial needs updating
The cert-manager tutorial needs a solid update:

- the version of cert-manager is very old and has an issue where it will generate excessive traffic to Let's Encrypt.
- the tutorial uses `j.hept.io` which will break soon if it hasn't already.
- The docs will also need updating after #950, #951, and #1140 are done. This is just to make sure that we don't miss these bits.

### PR DESCRIPTION
**Pull request by youngnick**
Fixes #1570
Fixes #1140
Updates #950
Updates #951

Signed-off-by: Nick Young <ynick@vmware.com>

### CODE REVIEW

File: site/_guides/cert-manager.md [hunk 1]
```diff
@@ -120,8 +118,9 @@ There are plenty of other ways to deploy cert-manager, but they are out of scope
 
 To keep things simple, we skip cert-manager's Helm installation, and use the supplied YAML manifests:
 
-```
-$ kubectl apply -f https://github.com/jetstack/cert-manager/releases/download/v0.8.0/cert-manager.yaml
+```sh
+$ kubectl create namespace cert-manager
+$ kubectl apply -f https://github.com/jetstack/cert-manager/releases/download/v0.11.0/cert-manager.yaml
```
Comment (davecheney) [inline site/_guides/cert-manager.md]: Does this command create the required namespace? It doesn't include a -n cert-manager flag, so it looks like the previous command isn't needed.

File: site/_guides/cert-manager.md [hunk 2]
```diff
@@ -164,10 +163,13 @@ metadata:
 spec:
   acme:
     email: user@example.com
-    http01: {}
     privateKeySecretRef:
-      name: letsencrypt-staging
+      name: letsencrypt-prod
```
Comment (davecheney) [inline site/_guides/cert-manager.md]: Is this change intention? The context of this file is the staging let's encrypt server

File: site/_guides/cert-manager.md [hunk 3]
```diff
@@ -363,27 +368,36 @@ spec:
 The `kubernetes.io/tls-acme: "true"` annotation tells cert-manager to use the `letsencrypt-staging` cluster-wide issuer we created earlier to request a certificate from Let's Encrypt's staging servers.
 
 The certificate is issued in the name of the hosts listed in the `tls:` section, `httpbin.davecheney.com` and stored in the secret `httpbin`.
-Behind the scenes, cert-manager creates a certificate CRD to manage the lifecycle of the certificate.
+Behind the scenes, cert-manager creates a certificate CRD to manage the lifecycle of the certificate, and then a series of other CRDs to handle the challenge process.
+The process goes `certificate` -> `certificaterequest` -> `order` -> `challenge`. If something goes wrong with this process, you will need to `kubectl get` then `kubectl describe` each of these objects in order to get the full picture.
```
Comment (davecheney) [inline site/_guides/cert-manager.md]: This background is useful, but incomplete. How would you feel about dropping this editorial?

File: site/_guides/cert-manager.md [hunk 4]
```diff
@@ -418,19 +432,23 @@ To request a properly signed certificate from the Let's Encrypt production serve
...
 spec:
   acme:
     email: user@example.com
-    http01: {}
     privateKeySecretRef:
       name: letsencrypt-prod
     server: https://acme-v02.api.letsencrypt.org/directory
+    solvers:
+    - http01:
+        ingress:
```
Comment (davecheney) [inline site/_guides/cert-manager.md]: Is the ingress class needed? What does it default to?

File: site/_guides/cert-manager.md [hunk 5]
```diff
@@ -487,11 +505,108 @@ $ curl https://httpbin.davecheney.com/get
...
+spec:
+  virtualhost:
+    fqdn: httpbinproxy.davecheney.com
+    tls:
+      secretName: httpbin
+  routes:
+  - services:
+    - name: httpbin
+      port: 8080
+
```
Comment (davecheney) [inline site/_guides/cert-manager.md]: s/blank line

File: site/_guides/cert-manager.md [hunk 6]
```diff
@@ -487,11 +505,108 @@ $ curl https://httpbin.davecheney.com/get
...
+Once that's done, create the dummy Ingress object:
+
+```yaml
+apiVersion: networking.k8s.io/v1beta1
+kind: Ingress
+metadata:
+  annotations:
+    cert-manager.io/cluster-issuer: letsencrypt-prod
+    ingress.kubernetes.io/force-ssl-redirect: "true"
+    kubernetes.io/ingress.allow-http: "false"
```
Comment (davecheney) [inline site/_guides/cert-manager.md]: I think ingress.allow-http defaults to false, we shouldn't need this annotation.

File: site/_guides/cert-manager.md [hunk 7]
```diff
@@ -487,11 +505,108 @@ $ curl https://httpbin.davecheney.com/get
...
+    kubernetes.io/ingress.class: contour
```
Comment (davecheney) [inline site/_guides/cert-manager.md]: why is this needed?

File: site/_guides/cert-manager.md [hunk 8]
```diff
@@ -487,11 +505,108 @@ $ curl https://httpbin.davecheney.com/get
...
+      - backend:
+          serviceName: httpbin
+          servicePort: 8080
+  tls:
+  - hosts:
+    - httpbinproxy.davecheney.com
+    secretName: httpbinproxy
+```
+
+Once cert-manager has done its thing, you will have a `kubernetes.io/tls` Secret present, that will contain the keypair.
```
Comment (davecheney) [inline site/_guides/cert-manager.md]: ... you will have a `httpbinproxy` secret.

File: site/_guides/cert-manager.md [hunk 9]
```diff
@@ -487,11 +505,108 @@ $ curl https://httpbin.davecheney.com/get
...
+  "url": "https://httpbinproxy.davecheney.com/get"
+}
+```
+
+### Caveats
+
+This method is a workaround until we can deliver the changes in [#950](https://github.com/projectcontour/contour/issues/950) and [#951](https://github.com/projectcontour/contour/issues/951).
+As such, it has a few caveats.
+
+Firstly, it works because Contour will override matching FQDN/path combinations in the following order: HTTPProxy overrides IngressRoute overrides Ingress.
```
Comment (davecheney) [inline site/_guides/cert-manager.md]: I don't think this is correct. You can overlay httpproxy on top of n Ingress records with the same fqdn. You cannot overlay httpproxy on top of IngressRoute to the best of my understanding.

File: site/_guides/cert-manager.md [hunk 10]
```diff
@@ -487,11 +505,108 @@ $ curl https://httpbin.davecheney.com/get
...
+The solver Ingress created by cert-manager has a unique FQDN/path combination so it will work, but the base one will be overridden by the HTTPProxy config.
```
Comment (davecheney) [inline site/_guides/cert-manager.md]: I don't think this is accurate. It might be simpler to delete this line.

File: site/_guides/cert-manager.md [hunk 11]
```diff
@@ -487,11 +505,108 @@ $ curl https://httpbin.davecheney.com/get
...
+
+Secondly, the actual config of where the object sends traffic is irrelevant, *as long as it matches a route in the HTTPProxy*.
```
Comment (davecheney) [inline site/_guides/cert-manager.md]: Would it be easier to say -- the dummy ingress record exists only to hold the annotations ingress-shim requires.

File: site/_guides/cert-manager.md [hunk 12]
```diff
@@ -487,11 +505,108 @@ $ curl https://httpbin.davecheney.com/get
...
+In this example, we're not using any advanced routing features, so this is straightforward.
```
Comment (davecheney) [inline site/_guides/cert-manager.md]: I don't think you need to document something that we're not doing. maybe just delete this line.

File: site/_guides/cert-manager.md [hunk 13]
```diff
@@ -487,11 +505,108 @@ $ curl https://httpbin.davecheney.com/get
...
+However, if you are, you'll either need to match routes between Ingress and HTTPProxy or create a dummy path for the Ingress to match on.
```
Comment (davecheney) [inline site/_guides/cert-manager.md]: I think this could be said simpler by saying the ingress record exists for ingress-shim only, mixing routes across ingress and httpproxy is outside the scope of this document.

File: site/_guides/cert-manager.md [hunk 14]
```diff
@@ -487,11 +505,108 @@ $ curl https://httpbin.davecheney.com/get
...
+
 ## Wrapping up
 
 Now that you've deployed your first HTTPS site using Contour and Let's Encrypt, deploying additional TLS enabled services is much simpler.
-Remember that for each HTTPS website you deploy, you create a Certificate CRD that provides the domain name and the name of the target Secret.
-After the Secret is created, you add a `tls` field to the Ingress object for your site.
+Remember that for each HTTPS website you deploy, cert-manager will  create a Certificate CRD that provides the domain name and the name of the target Secret.
+After the Secret is created, you add a `tls` field to the HTTPProxy or Ingress object for your site.
```
Comment (davecheney) [inline site/_guides/cert-manager.md]: The order of operations isn't this strict, contour is just not going to be able to serve the https version of the site until the secret is ready.

File: site/_guides/cert-manager.md (hunk 1 above)
Comment (jpeach) [inline site/_guides/cert-manager.md]: Note the need to pass `--validate=false` here for Kubernetes < 1.15

File: site/_guides/cert-manager.md [hunk 15]
```diff
@@ -487,11 +505,108 @@ $ curl https://httpbin.davecheney.com/get
...
+    - name: httpbin
+      port: 8080
+
+```
+
+This config will be marked as Invalid by Contour, since the TLS secret doesn't exist yet.
+Once that's done, create the dummy Ingress object:
+
+```yaml
+apiVersion: networking.k8s.io/v1beta1
```
Comment (jpeach) [inline site/_guides/cert-manager.md]: Earlier, the YAML uses `extensions/v1beta1` for Ingress. We should be consistent across the whole page and note the k8s version that require `extensions/`.

File: site/_guides/cert-manager.md [hunk 16]
```diff
@@ -103,7 +101,7 @@ kuard-67ff6dd458-sfxkb     1/1       Running   0          19d
 
 Then type the DNS name you set up in the previous step into a web browser, for example `http://gke.davecheney.com/`. You should see something like:
 
-![KAURD screenshot][kuard]
+![KUARD screenshot][kuard]
```
[the remark this answers is not in the document]
Comment (youngnick) [reply site/_guides/cert-manager.md]: I guess, sure.

File: site/_guides/cert-manager.md (hunk 3 above)
Comment (youngnick) [reply site/_guides/cert-manager.md]: I'm okay with it. I'll remove.

File: site/_guides/cert-manager.md (hunk 4 above)
Comment (youngnick) [reply site/_guides/cert-manager.md]: There is no default. You must specify this for cert-manager to know the ingress class to put on the Ingress objects it creates. This is new in 0.11.

File: site/_guides/cert-manager.md (hunk 15 above)
Comment (youngnick) [reply site/_guides/cert-manager.md]: Good point, will do.

File: site/_guides/cert-manager.md (hunk 7 above)
Comment (youngnick) [reply site/_guides/cert-manager.md]: hmm, good point. I'll check if it is needed.

File: site/_guides/cert-manager.md (hunk 11 above)
Comment (youngnick) [reply site/_guides/cert-manager.md]: Yep, that's better, thanks.

File: site/_guides/cert-manager.md (hunk 14 above)
Comment (youngnick) [reply site/_guides/cert-manager.md]: Okay, I'll write it differently, thanks.

File: site/_guides/cert-manager.md (hunk 1 above)
Comment (youngnick) [reply site/_guides/cert-manager.md]: Yeah, I suspect this is to allow the management of the `kubectl delete` flow.

Not deleting the namespace as part of the YAML means that you can `kubectl delete -f <file>` and you won't get stuck with the namespace not terminating because of the ValidatingWebhook config.

File: site/_guides/cert-manager.md (hunk 2 above)
Comment (youngnick) [reply site/_guides/cert-manager.md]: Nope, will revert.

File: site/_guides/cert-manager.md (hunk 7 above)
Comment (youngnick) [reply site/_guides/cert-manager.md]: It is not needed, which simplifies this quite a bit.

**Comment by youngnick** [comment]
This is ready for rereview.

File: site/_guides/cert-manager.md [hunk 17]
```diff
@@ -85,66 +83,73 @@ In your own data center, you need to arrange for traffic from a public IP addres
...
 ```bash
-$ kubectl delete -f {{ site.url }}/examples/kuard.yaml
+$ kubectl delete -f {{ site.url }}/examples/httpbin.yaml
 ```
 
 ## 2. Deploy jetstack/cert-manager
 
 **NOTE:** cert-manager is a powerful product that provides more functionality than this tutorial demonstrates.
-There are plenty of other ways to deploy cert-manager, but they are out of scope.
+There are plenty of [other ways to deploy cert-manager](4), but they are out of scope.
```
[the remark this answers is not in the document]
Comment (youngnick) [reply site/_guides/cert-manager.md]: 🤦‍♂ thanks
